EQ optimazations

BaraMGB opened the pull request:
1. The Curves will be only calculate when the data has changed.
2. The Spectrum Analyzer calculates only if a there is a signal.
3. If there is no signal anymore, the graph is going to zero now.

jasp00 commented on plugins/Eq/EqParameterWidget.cpp:
```diff
@@ -92,6 +92,7 @@ EqParameterWidget::~EqParameterWidget()
 
 void EqParameterWidget::updateView()
```
The `updateView()` name is unfortunate because the method changes the model, but this is not introduced by this request.

jasp00 commented on plugins/Eq/EqSpectrumView.h:
```diff
@@ -25,8 +25,10 @@
+
 #include "Engine.h"
+#include "fft_helpers.h"
+#include "GuiApplication.h"
```
Is this needed?

jasp00 commented on plugins/Eq/EqSpectrumView.h:
```diff
@@ -45,27 +47,43 @@ class EqAnalyser
+
 
 
 	EqAnalyser() :
```
I fail to see why there is no `EqSpectrumView.cpp` with all these implementations.

jasp00 commented on plugins/Eq/EqSpectrumView.h:
```diff
@@ -177,23 +196,33 @@ class EqSpectrumView : public QWidget
 		QPainter p( this );
 		p.setPen( QPen( color, 1, Qt::SolidLine, Qt::RoundCap, Qt::BevelJoin ) );
-		const float e =  m_sa->m_energy;
+		float e =  m_sa->m_energy;
```
Why do you remove `const`?

jasp00 commented on plugins/Eq/EqSpectrumView.h:
```diff
@@ -202,16 +231,29 @@ class EqSpectrumView : public QWidget
+			}
+
+			pp.lineTo( freqToXPixel( bandToFreq( x ) ), fh - m_bandHeight[x] );
+			h_sum = h_sum + m_bandHeight[x];
```
You may like `h_sum += m_bandHeight[x];`.

jasp00 commented on plugins/Eq/EqSpectrumView.h:
```diff
@@ -202,16 +231,29 @@ class EqSpectrumView : public QWidget
+
+		if( h_sum == 0 )
+		{
+			m_sa->m_energy = 0;
```
This code should not be in a paint event. This should be handled in `analyze()`.

BaraMGB commented:
@jasp00 I updated this.

- spectrum analyzer performance improved
- new file EqSpectrumAnalyser.cpp
- avoiding high cpu usage on painting QPainterPaths through cache in a QPixmap
- a lot of code changes according to the coding guideline
- cleaned up the header files. 
- a lot of member variables removed when not needed

BaraMGB commented:
Okay, I connected the spectrum view update to the gui->mainWindow()->periodicUpdate() and implemented the calculate x, y from/to gain, frequency functions only once (don't repeat yourself).

After someone reviewd the code it can be merged. ( @jasp00 ?)

jasp00 commented:
I doubt I can work on LMMS until weekend and this request is not trivial.

jasp00 commented on plugins/Eq/EqSpectrumView.cpp:
```diff
@@ -0,0 +1,294 @@
+
+
+
+float EqSpectrumView::freqToXPixel(float freq)
```
`EqHandle::freqToXPixel()` could be reused.

BaraMGB commented:
@jasp00 thank you for reviewing this. I changed the suggestions of you. 
If you want to test this, I suggest to compare with the master branch. Please, have an eye on the system CPU usage. 

- Especially on idle with all bands active. 
- The spectrum analyzer should only calculate (and paint!) if a signal is going in.

best regards

jasp00 commented on plugins/Eq/EqCurve.cpp:
```diff
@@ -304,37 +319,18 @@ float EqHandle::getLowCutCurve( float x )
-						  + ( b0 * b2 * PHI - ( b1 * ( b0 + b2 ) + 4 * b0 * b2 ) ) * PHI )
-						  - 10 * log10( pow( 1 + a1 + a2, 2 ) + ( 1 * a2 * PHI - ( a1 * ( 1 + a2 ) + 4 * 1 * a2 ) ) * PHI ));
-
+		gain = gain * 4;
```
Are you sure it is `gain * 4`? Two sums means `gain * 3`. On the other side, regarding `m_hp48` 48 is twice 24 (`gain * 2`); are you changing it because of this?

BaraMGB replied:
You are right. It's not linear.

jasp00 commented:
The patch reduces CPU usage. The graph does not go to zero if the analyzer is disabled, but this behavior was already present and can be fixed later. This request is an improvement; it should be merged after considering https://github.com/LMMS/lmms/pull/3036#discussion_r82531097.

jasp00 commented on plugins/Eq/EqCurve.cpp:
```diff
@@ -304,37 +319,18 @@ float EqHandle::getLowCutCurve( float x )
-						  + ( b0 * b2 * PHI - ( b1 * ( b0 + b2 ) + 4 * b0 * b2 ) ) * PHI )
-						  - 10 * log10( pow( 1 + a1 + a2, 2 ) + ( 1 * a2 * PHI - ( a1 * ( 1 + a2 ) + 4 * 1 * a2 ) ) * PHI ));
-
+        gain = gain * 3;
```
Indentation with spaces?

jasp00 commented on plugins/Eq/EqCurve.cpp:
```diff
@@ -369,45 +363,24 @@ float EqHandle::getHighCutCurve( float x )
-		gain = gain + ( 10 * log10( pow( b0 + b1 + b2 , 2 )
-						  + ( b0 * b2 * PHI - ( b1 * ( b0 + b2 ) + 4 * b0 * b2 ) ) * PHI )
-						  - 10 * log10( pow( 1 + a1 + a2, 2 ) + ( 1 * a2 * PHI - ( a1 * ( 1 + a2 ) + 4 * 1 * a2 ) ) * PHI ) );
+        gain = gain * 3;
```
And this one.

BaraMGB replied:
corrected.

BaraMGB commented:
Thank you for your time.
